### PR TITLE
feat(eslint-plugin): [no-invalid-void-type] better report message for void used as a constituent inside a function return type 

### LINKED ISSUE #5255: Bug: [no-invalid-void-type] better report message for void used as a constituent inside a function return type
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwAoBKALnlWQFsAjEGeAH3gDcctgBuAWACggA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6KAHc+0JpHBgAviDVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### PR DESCRIPTION
**Pull request by kmin-jeong**
## PR Checklist

-   [x] Addresses an existing open issue: fixes #5255
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview
change and  error message in case of union type

### CODE REVIEW

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Thanks for starting this PR! Let's talk about strategy.

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts [hunk 1]
```diff
@@ -25,7 +25,7 @@ export default util.createRule<[Options], MessageIds>({
       invalidVoidForGeneric:
         '{{ generic }} may not have void as a type variable.',
       invalidVoidNotReturnOrGeneric:
-        'void is only valid as a return type or generic type variable.',
+        'void is valid as a return type in the case of a union or intersection type and generic type variable',
```
[the remark this answers is not in the document]
Comment (kmin-jeong) [reply packages/eslint-plugin/src/rules/no-invalid-void-type.ts]: @JoshuaKGoldberg 
I separate error message about void in union type used as a return type!
Is intersection type  to be considered?

**Comment by kmin-jeong** [comment]
I separate error messages named message variables 'InvalidVoidForUnion', and put 'void is only valid as a return type in the case of a union'.  I'm looking for a way to squash for the records that were committed to past pr that I solved in the past.
Is there anything else to consider about strategy?

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts [hunk 2]
```diff
@@ -31,6 +32,8 @@ export default util.createRule<[Options], MessageIds>({
         'void is only valid as return type or type of `this` parameter.',
       invalidVoidNotReturnOrThisParamOrGeneric:
         'void is only valid as a return type or generic type variable or the type of a `this` parameter.',
+      InvalidVoidForUnion:
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-invalid-void-type.ts:35]: This new `InvalidVoidForUnion` isn't used anywhere...?

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts [hunk 3]
```diff
@@ -31,6 +32,8 @@ export default util.createRule<[Options], MessageIds>({
...
+        'Void is valid when any types used together only as part of a union type.',
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-invalid-void-type.ts:36]: This sentence doesn't make sense to me. Also, `void` is lowercase in the other messages. How about...

```suggestion
        'void is not valid as a type constituent in a return type.',
```

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Sorry for the 2-3 weeks of silence, was out but back in reviewing mode now! I think something got lost in the latest commits as the `InvalidVoidForUnion` message isn't used anywhere?

For what it's worth, we don't care about the Git history within a PR being clean. The commit will be squash merged anyway.

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts (hunk 3 above)
Comment (kmin-jeong) [reply packages/eslint-plugin/src/rules/no-invalid-void-type.ts]: ok! I fixed this error message!

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts (hunk 2 above)
Comment (kmin-jeong) [reply packages/eslint-plugin/src/rules/no-invalid-void-type.ts:36]: Except for the error message, I didn't touch the rest  like testcode.

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/no-invalid-void-type.ts:36]: Yes, that's the issue 😄. Defining a message on its own doesn't do anything. You have to modify the rule's logic to report that message too.

**Comment by JoshuaKGoldberg** [comment]
👋 Just checking in @kmin-jeong, is this PR still something you have the time to work on? No worries if not, just making sure the issue isn't blocked. 😄

**Comment by kmin-jeong** [comment]
@JoshuaKGoldberg 
Is there anything I need to check so that the issue doesn't close?

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts (hunk 2 above)
Comment (kmin-jeong) [reply packages/eslint-plugin/src/rules/no-invalid-void-type.ts:37]: @JoshuaKGoldberg 
I modified the test case and logic, but I'm not sure if this is right

**Comment by JoshuaKGoldberg** [comment]
> Is there anything I need to check so that the issue doesn't close?

Nope, PRs generally stay open unless some drastic difference in strategy is decided upon (not likely here) or they aren't updated by the author in months.

File: packages/eslint-plugin/tests/rules/no-invalid-void-type.test.ts [hunk 4]
```diff
@@ -98,6 +98,28 @@ function foo(): void | never {
   ],
 });
 
+ruleTester.run('invalidUnionParameter:true', rule, {
+  valid: [
+    {
+      code: 'type allowUnion = string | number',
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-invalid-void-type.test.ts:104]: @kmin-jeong this PR has some failures. GitHub shows some of them in the web UI (example: the one right here!). You can find the rest of them in the checks view on the main PR page:

image

In particular, we can see tests failing: https://github.com/typescript-eslint/typescript-eslint/runs/7897893084?check_suite_focus=true

You can run just the tests for the `no-invalid-void-type` rule by `cd`ing into `packages/eslint-plugin` and running `npx jest no-invalid-void-type`:

```shell
cd packages/eslint-plugin
npx jest no-invalid-void-type
```

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts [hunk 5]
```diff
@@ -49,15 +54,33 @@ export default util.createRule<[Options], MessageIds>({
...
-    { allowInGenericTypeArguments: true, allowAsThisParameter: false },
+    {
+      allowInGenericTypeArguments: true,
+      allowAsThisParameter: false,
+      invalidUnionParameter: true,
+      allowUnionType:true,
+    },
   ],
-  create(context, [{ allowInGenericTypeArguments, allowAsThisParameter }]) {
+  create(context, [{ allowInGenericTypeArguments, allowAsThisParameter, invalidUnionParameter,allowUnionType }]) {
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-invalid-void-type.ts:83]: May I suggest installing the [Prettier](https://prettier.io) extension for your editor? 😄

**Review by JoshuaKGoldberg — DISMISSED** [review]
Let me know if you need help with any of this 🙁. I get the impression this is a lot of new stuff - feel free to ask questions on anything that doesn't make sense!

https://sourcegraph.com/notebooks/Tm90ZWJvb2s6MTA2OA== has some good docs on how our rules work, if that helps.

**Comment by JoshuaKGoldberg** [comment]
Ping @kmin-jeong, do you still have time to work on this? No worries if not, just want to make sure it's not pending too long. 🙂

**Comment by kmin-jeong** [comment]
@JoshuaKGoldberg I don't have enough time to solve this , so can I hand it over to someone else?

**Review by bradzacher — APPROVED** [review]

**Pull request merged**